File: cpp/recursion/head_tail_recursion.cpp
#include <iostream>

using namespace std;


/*
Head recursion -> when there is no operation done before function call or
function call is the first statement of recursive function block.

All the other operation execute after the function calls.

It means function does not perform anything before calling but it execute everything 
at returning time.

*/
void fun2(int n)
{
    if (n > 0)
    {
        fun2(n-1); 
        printf("%d ",n);
    }
}

int main()
{
    int x = 10;
    fun2(x);
    return 0;
}

File: python/recursion/head_tail_recursion.py
# tail recursion
# -> ascending output -> print at calling time
def fun1(n):
    if n>0:
        print(n, end=' ')
        fun1(n-1)

# head recursion
# descending output -> print at returning time

def fun2(n):
    if n>0:
        fun2(n-1)
        print(n, end=' ')

n = int(input())
fun1(n)
fun2(n)